pkg/fswatcher/fswatcher.go

```diff
@@ -133,12 +139,21 @@ func (w *FSWatcher) watch() {
 				return
 			}
 			w.logger.Error("fsnotifier reported an error", zap.Error(err))
+		case <-w.stop:
+			w.wg.Done()
+			return
 		}
 	}
 }
 
-// Close closes the watcher.
+// Close closes the watcher and stops the background go routine of FSWatcher.
 func (w *FSWatcher) Close() error {
+	if !w.isClosed {
+		close(w.stop)
+		w.wg.Wait()
+	}
+	w.isClosed = true
```

Comment:
unprotected bool can be a source of race condition.

Reply:
I see, let me add the mutex lock to prevent the race conditions.